linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.interface
+
+/** A single Module output specification for a linker run.
+ *
+ *  The fields [[moduleID]] and [[isInternal]] uniquely identify this output in
+ *  a [[Linker]] run.
+ *
+ *  @param moduleID The ID of the module this linker output represents.
+ *
+ *  @param isInternal Whether this is an internal (or an external) module.
+ *
+ *  @param output The [[LinkerOutput]] for this module.
+ *
+ *  @param moduleName Name of this module for import. A [[Linker]] will use
+ *      this name to include this module from one of its dependees. Must be set,
+ *      if this module is internal.
+ */
+final class ModuleLinkerOutput private (
+    val moduleID: ModuleOutputSpec.ModuleID,
+    val isInternal: Boolean,
+    val output: LinkerOutput,
+    val moduleName: Option[String]
+) {
+  private def this(moduleID: ModuleOutputSpec.ModuleID, output: LinkerOutput) =
+    this(moduleID, false, output, None)
+
+  def withInternal: ModuleLinkerOutput =
```

Comment:
If an internal module must have a module name, perhaps `withInternal` should take `moduleName` as parameter?

Reply:
Probably a better design choice indeed. Although see my previous comment about `moduleName` being meaningful all the time and this maybe being unnecessary burden on upstream code.

Reply:
I'm not sure it helps unless `withInternal` takes a boolean. Because it's a choice between (currently):
```scala
val moduleOutput0 = ModuleLinkerOutput(???)
  .withModuleName(someModuleName)
val moduleOutput =
  if (isInternal) moduleOutput0.withInternal
  else moduleOutput0
```
and (with `withModule` taking the `moduleName`):
```scala
val moduleOutput0 = ModuleLinkerOutput(???)
val moduleOutput =
  if (isInternal) moduleOutput0.withInternal(someModuleName)
  else moduleOutput0
```
It would be simplified if `withInternal` took a boolean instead, as that would allow:
```scala
val moduleOutput = ModuleLinkerOutput(???)
  .withModuleName(someModuleName)
  .withInternal(isInternal)
```